### PR TITLE
Deduplicate file names without ordering them

### PR DESCRIPTION
**Pull request by Morriar**
While fixing issue #203, the PR #755 removed the ability for the user to control the file loading order.
This PR accomplish the same thing than #755 while keeping them as specified by the user.

### Motivation

Be the following RBI file in a `foo/` directory:

~~~ruby
# foo/stdlib.rbi
# typed: true

class A
end

class Foo
  extend T::Sig

  sig { params(a: A).void }
  def foo(a); end
end
~~~

And Ruby a module using it:

~~~ruby
# foo/foo.rb
# typed: true

class Foo
  def foo(a, b); end
end
~~~

Because of the file names, when passing `foo/` to Sorbet, we will get this error message:

~~~bash
# Loading `foo` only gives the bad order
$ sorbet foo/

stdlib.rbi:10: Method Foo#foo redefined without matching argument count. Expected: 2, got: 1
    10 |  def foo(a); end
          ^^^^^^^^^^
    foo.rb:4: Previous definition
     4 |  def foo(a, b); end
          ^^^^^^^^^^^^^
Errors: 1
~~~

Notice that `stdlib.rbi` is wrongly presented in the error as the redefinition from `foo.rb`.

Before #755 (commit db452936f1b2c91301503b2442ddd72b7a94616f), it was possible to go around this by specifying the order of the arguments:

~~~bash
# Changing the arguments order works
$ sorbet foo/stdlib.rbi foo/foo.rb

foo/foo.rb:4: Method Foo#foo redefined without matching argument count. Expected: 1, got: 2
     4 |  def foo(a, b); end
          ^^^^^^^^^^^^^
    foo/stdlib.rbi:10: Previous definition
    10 |  def foo(a); end
          ^^^^^^^^^^
Errors: 1
~~~

After #755 (commit 0fc29b4c38a01f9663a14a1ccfb89f883373938a), the fix for #203 removed this ability:

~~~bash
# Changing the order doesn't work anymore
$ sorbet foo/stdlib.rbi foo/foo.rb

foo/stdlib.rbi:10: Method Foo#foo redefined without matching argument count. Expected: 2, got: 1
    10 |  def foo(a); end
          ^^^^^^^^^^
    foo/foo.rb:4: Previous definition
     4 |  def foo(a, b); end
          ^^^^^^^^^^^^^
Errors: 1
~~~

This PR (commit 2097f831fa72fff09c50ebebc2043e84fb5c5cd6) adds it back:

~~~bash
# Changing the arguments order works again
$ sorbet foo/stdlib.rbi foo/foo.rb

foo/foo.rb:4: Method Foo#foo redefined without matching argument count. Expected: 1, got: 2
     4 |  def foo(a, b); end
          ^^^^^^^^^^^^^
    foo/stdlib.rbi:10: Previous definition
    10 |  def foo(a); end
          ^^^^^^^^^^
Errors: 1
~~~

### Benchmarks

I benchmarked the solution by loading the content of `rbi/` with the following command:

~~~bash
find rbi -name "*.rbi" | grep -v light.rbi > benchfiles
bazel-bin/main/sorbet --no-stdlib @benchfiles
~~~

The following table gives the average loading time over 10 runs sor each version of Sorbet:

| desc.			| commit									| time (s)	|
|---------------|-------------------------------------------|-----------|
| before-755	| db452936f1b2c91301503b2442ddd72b7a94616f	| 2.280		|
| after-755	| 0fc29b4c38a01f9663a14a1ccfb89f883373938a	| 2.143		|
| this-pr		| 2097f831fa72fff09c50ebebc2043e84fb5c5cd6	| 2.118		|

### Test plan

I added a tests showing the behaviour before/after on dummy files as well as real ones, directories and config files.

### CODE REVIEW

**Comment by Morriar** [comment]
Rebased to remove conflicts with `master`.

File: main/options/options.cc [hunk 1]
```diff
@@ -481,21 +481,34 @@ void Options::flushPrinters() {
...
+void addFile(Options &opts, UnorderedSet<string> &knownFileNames, string file) {
+    if (knownFileNames.find(file) == knownFileNames.end()) {
+        knownFileNames.insert(file);
+        opts.inputFileNames.push_back(file);
+    }
+}
+
+void addFiles(Options &opts, UnorderedSet<string> &knownFileNames, vector<string> files) {
+    opts.inputFileNames.reserve(opts.inputFileNames.size() + files.size());
+    for (auto file : files) {
```
Comment (XrXr) [inline main/options/options.cc]:
```suggestion
    for (const auto &file : files) {
```

File: main/options/options.cc [hunk 2]
```diff
@@ -481,21 +481,34 @@ void Options::flushPrinters() {
     }
 }
 
-void addFilesFromDir(Options &opts, string_view dir) {
+void addFile(Options &opts, UnorderedSet<string> &knownFileNames, string file) {
```
Comment (XrXr) [inline main/options/options.cc]:
```suggestion
void addFile(Options &opts, UnorderedSet<string> &knownFileNames, const string &file) {
```

Comment (DarkDimius) [reply main/options/options.cc]: can the `file` argument be a `string_view`? We prefer them as 'non-owned` strings

**Comment by Morriar** [comment]
@XrXr: thanks for the suggestions!

@DarkDimius: changed `string` by `string_view`.

@jez:

> For example, we might want to make this error independent of the file order, and consistently report the error in one location.

Agreed, Sorbet lacks some specification on that point.

Maybe `rbi` files should always be loaded first?

> We might also want to just hasten our efforts to consolidate on one source of truth for the stdlib RBIs, so they're always ordered consistently because they're coming from the same place.

Forcing people to use `sorbet-typed` (or whatever the source of truth will be) seems pretty harsh to solve this problem. What if I don't want to load all the RBIs provided? Should I be creative with my filenames so it's loaded in the right order?

I would suggest merging this PR as it solves the problem (at list for sorbet, not LSP) today and make us able to use Sorbet again then working on Sorbet specification so we have a clean way to order definitions.

**Comment by DarkDimius** [comment]
@Morriar I'm there with @jez, we're trying to see how to make Sorbet as much order independent as possible for better UX and for our own sanity: it's hard to debug a program that doesn't work reproducibly.

From this point of view, stabilizing order of input files seems like a boon to me and I'd like to keep it.

I'd welcome making this "redefinition error" smarter, would this address the issue you're having?

**Comment by Morriar** [comment]
Alright, let's go with ordering the files by their names.

**Pull request closed without merging**